pyro/infer/contract.py

```diff
@@ -126,6 +126,8 @@ def _contract_component(tensor_tree, sum_dims, target_ordinal=None):
     for t, terms in tensor_tree.items():
         dims_tree[t] = set.union(set(), *(sum_dims[term] for term in terms))
 
+    enum_boundary = max(set.union({float("-inf")}, *sum_dims.values()))
```

Comment:
I'm not sure this is the right level of abstraction for this function, especially there is no mention of enumeration in this file. Do you think it's possible to move this up the stack a bit? Will this be fixed by `ring.broadcast()` in #1405? What was the error that this fixed?

Reply:
Also, just checked that this still errors on #1405 - the failing test is `tests.infer.mcmc.test_valid_models.test_all_discrete_sites_log_prob`.

Reply:
Hmm it would be good to have a failing test in `test/infer/test_contract.py`. I'll try adding one, assuming the cause is `frame.size == 1`.

Note the original motivation for aggressively squeezing was to maximize sharing via `shared_intermediates`. My mental model is not precise enough to gauge how much extra sharing we get.

Reply:
Great, it works now! I'll take another look at #1405.